### DISCUSSION TITLE
Invariant Test failing, fuzz error

### DISCUSSION THREAD

**Discussion by YavorJJ** [Cyfrin/foundry-full-course-cu#2008] [Q&A]
I am receiving this error when I test:

` ` `
[FAIL. Reason: failed to set up invariant testing environment: No contracts to fuzz.] invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 0, calls: 0, reverts: 0)

` ` `

Here is my code: 

` ` ` solidity 
         // SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// Have our Invariant aka properties

// What are our invariants ?

// 1. The total supply of DSC should be less than the total value of collateral
// 2. Getter view functions should never revert <- evergreen invariant

import {Test, console} from "forge-std/Test.sol";
import {StdInvariant} from "forge-std/StdInvariant.sol";
import {DeployDSC} from "../../script/DeployDSC.s.sol";
import {DSCEngine} from "../../src/DSCEngine.sol";
import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";

contract OpenInvariantsTest is StdInvariant, Test {
    DeployDSC deployer;
    DSCEngine dsce;
    DecentralizedStableCoin dsc;
    HelperConfig config;
    address weth;
    address wbtc;

    function setUp() external{
        deployer = new DeployDSC();
        (dsc, dsce, config) = deployer.run(); 
        (,,weth, wbtc,) = config.activeNetworkconfig();
        targetContract(address(dsce));
    }

    function invariant_protocolMustHaveMoreValueThanTotalSupply() public view{
        // get the value of all the collateral in the protocol
        // compare it to all the debt (dsc)
        uint256 totalSupply = dsc.totalSupply();
        uint256 totalWethDeposited = IERC20(weth).balanceOf(address(dsce));
        uint256 totalWbtcDeposited = IERC20(wbtc).balanceOf(address(dsce));
        uint256 wethValue = dsce.getUsdValue(weth, totalWethDeposited);
        uint256 wbtcValue = dsce.getUsdValue(wbtc, totalWbtcDeposited);

        assert(wbtcValue + wethValue > totalSupply);
    }
}
` ` `
Can someone help solving and fixing this problem.       
Thank you.

**Comment by EngrPips** [comment]
Hello @YavorJJ, Please run `foundryup` then retry. Let's see how it goes.

**Reply by YavorJJ** [reply]
just did, it stays the same: 
![image](https://github.com/user-attachments/assets/22105423-e7ed-4349-bb33-822434034fa9)

**Reply by EngrPips** [reply]
Please show your deployer contract.

**Reply by YavorJJ** [reply]
> Please show your deployer contract.

sure.

` ` `solidity
// SPDX-Lincense-Identifier: MIT

pragma solidity ^0.8.19;

import {Script} from "forge-std/Script.sol";
import {DecentralizedStableCoin} from "../src/DecentralizedStableCoin.sol";
import {DSCEngine} from "../src/DSCEngine.sol";
import {HelperConfig} from "./HelperConfig.s.sol";

contract DeployDSC is Script {
    address[] public tokenAddresses;
    address[] public priceFeedAddresses;


    function run() external returns (DecentralizedStableCoin, DSCEngine, HelperConfig) {
        HelperConfig config = new HelperConfig();

        (address wethUsdPriceFeed, address wbtcUsdPriceFeed, address weth, address wbtc, uint256 deployerKey) = config.activeNetworkconfig();

        tokenAddresses = [weth, wbtc];
        priceFeedAddresses = [wethUsdPriceFeed, wbtcUsdPriceFeed];
     
        vm.startBroadcast(deployerKey);
        DecentralizedStableCoin dsc = new DecentralizedStableCoin();
        DSCEngine engine = new DSCEngine(tokenAddresses, priceFeedAddresses, address(dsc));

        dsc.transferOwnership(address(engine));
        vm.stopBroadcast();
        return (dsc, engine, config);
    }
}   

` ` `

**Comment by EngrPips** [comment]
```js
// SPDX-Lincense-Identifier: MIT

pragma solidity ^0.8.19;

import {Script} from "forge-std/Script.sol";
import {DecentralizedStableCoin} from "../src/DecentralizedStableCoin.sol";
import {DSCEngine} from "../src/DSCEngine.sol";
import {HelperConfig} from "./HelperConfig.s.sol";

contract DeployDSC is Script {
address[] public tokenAddresses;
address[] public priceFeedAddresses;

function run() external returns (DecentralizedStableCoin, DSCEngine, HelperConfig) {
    HelperConfig config = new HelperConfig();

    (address wethUsdPriceFeed, address wbtcUsdPriceFeed, address weth, address wbtc, uint256 deployerKey) = config.activeNetworkconfig();

    tokenAddresses = [weth, wbtc];
    priceFeedAddresses = [wethUsdPriceFeed, wbtcUsdPriceFeed];
 
    vm.startBroadcast(deployerKey);
    DecentralizedStableCoin dsc = new DecentralizedStableCoin();
    DSCEngine engine = new DSCEngine(tokenAddresses, priceFeedAddresses, address(dsc));

    dsc.transferOwnership(address(engine));
    vm.stopBroadcast();
    return (dsc, engine, config);
}
}
```

**Comment by EngrPips** [comment]
Hello @YavorJJ, it seems like you did everything correctly, and I am unsure why you are getting that error. Please push the code to Github I will clone and test on my machine to see if I will catch what is wrong.

**Reply by YavorJJ** [reply]
I did a research for this, and i dont know even if deploying it to github will fix this, some said that sometimes Patrick's foundry and testing acts a bit different from ours.

**Reply by EngrPips** [reply]
Interesting. It is completely normal that when we are following a course, some things would work for the instructor and not work for us and that might be because of an update in the tools used or a silly mistake made by us that we can't catch. So we always have to take it as a task upon ourselves to make things work for us

**Reply by YavorJJ** [reply]
> Interesting. It is completely normal that when we are following a course, some things would work for the instructor and not work for us and that might be because of an update in the tools used or a silly mistake made by us that we can't catch. So we always have to take it as a task upon ourselves to make things work for us

it is up on my github, would be gratefull to check. and yes of course like you said we should always check, it could be a stupid mistake that ruins my code and test.

**Reply by EngrPips** [reply]
All right, I will do so as soon as I can.

**Reply by YavorJJ** [reply]
> All right, I will do so as soon as I can.

thanks, much appreciated

**Reply by EngrPips** [reply]
You are welcome.

**Reply by YavorJJ** [reply]
any update bro? I still cant figure it out here :D

**Reply by YavorJJ** [reply]
> You are welcome.

so, I did forge clean and forge build, then forge test --mt invariant_protocolMustHaveMoreValueThanTotalSupply() and suddenly I get this error. 
![image](https://github.com/user-attachments/assets/51b7a34a-d9e3-4b3c-85a5-a880301b1d27)

I guess now its good that there is a contract to fuzz, but still there is an error.

**Reply by EngrPips** [reply]
`Assertion failed` means you made some mistake in your contract, which makes your assertion wrong. You should copy the output of the test and paste it on here next time instead of sending pictures. That will help us see everything that happened and we might figure out what might be wrong.

**Reply by YavorJJ** [reply]
> `Assertion failed` means you made some mistake in your contract, which makes your assertion wrong. You should copy the output of the test and paste it on here next time instead of sending pictures. That will help us see everything that happened and we might figure out what might be wrong.

yep, figured it out i messed up the equal sign.

**Reply by EngrPips** [reply]
Awesome, I am glad you figured it out. Keep crushing it, sir.

**Reply by nayashi002** [reply]
thanks was getting the same error now but was resolved using procedures

**Reply by EngrPips** [reply]
Interesting, Keep crushing it.

**Reply by BTBMan** [reply]
> > You are welcome.
> 
> so, I did forge clean and forge build, then forge test --mt invariant_protocolMustHaveMoreValueThanTotalSupply() and suddenly I get this error. ![image](https://private-user-images.githubusercontent.com/171678195/348542698-51b7a34a-d9e3-4b3c-85a5-a880301b1d27.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgzNjI0NjYsIm5iZiI6MTcyODM2MjE2NiwicGF0aCI6Ii8xNzE2NzgxOTUvMzQ4NTQyNjk4LTUxYjdhMzRhLWQ5ZTMtNGIzYy04NWE1LWE4ODAzMDFiMWQyNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwOFQwNDM2MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hY2JiYzYzOWM5YjhkOGYzZDM4OWQ1NzE5ODczYjJkM2I1Y2E1YmVmYmEwZjlhYzY0NmEzNmZmYzk4NTI4M2E3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4aSgS_Nuz5ExmspdowOhIpZgOAYuslrY82Q1JNPz8nw)
> 
> I guess now its good that there is a contract to fuzz, but still there is an error.

`forge clean` works for me, Thanks.

**Reply by EngrPips** [reply]
Did you delete your `cache` folder too

**Reply by BTBMan** [reply]
No, I just run `forge cache`, It seems always have cache problem with invariant testing, So do `forge cache` any time or `forge test --force`.😂

**Reply by EngrPips** [reply]
Yeah the caching have been a problem lately.

**Reply by rishu605** [reply]
This helped:
`forge clean`
`forge test`

**Reply by EngrPips** [reply]
Awesome.

**Comment by JoshuyPan** [comment]
Guys, i'm having the same issue and i can't figure it out. I also tried asking IABuddy, nothing usefull from it.
Here is my code:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// have our invatiants aka properties

//What are our invatiants?

// 1. The total supply og DSC should be less tha the total value of collateral
// 2. Getter view functions should never revert <- evergreen invariant
import {Test, console} from "forge-std/Test.sol";
import {StdInvariant} from "forge-std/StdInvariant.sol";
import {DeployDSC} from "script/DeployDSC.s.sol";
import {DSCEngine} from "src/DSCEngine.sol";
import {DecentralizedStableCoin} from "src/DecentralizedStableCoin.sol";
import {HelperConfig} from "script/HelperConfig.s.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";

contract InvaritantTest is StdInvariant, Test {
    DeployDSC deployer;
    DSCEngine dsce;
    DecentralizedStableCoin dsc;
    HelperConfig config;
    address weth;
    address wbtc;

    function setUp() external {
        deployer = new DeployDSC();
        (dsc, dsce, config) = deployer.run();
        (,, weth, wbtc,) = config.activeNetworkConfig();
        targetContract(address(dsce));
    }

    function invariant_protocolMustHaveMoreValueThanTotalSupply() public view {
        // get the value of all the collateral in the protocol
        // compare it to all the debt (dsc)
        uint256 totalSupply = dsc.totalSupply();
        uint256 totalWethDeposited = IERC20(weth).balanceOf(address(dsce));
        uint256 totalWbtcDeposited = IERC20(wbtc).balanceOf(address(dsce));

        uint256 wethValue = dsce.getUsdValue(weth, totalWethDeposited);
        uint256 wbtcValue = dsce.getUsdValue(wbtc, totalWbtcDeposited);

        console.log("totalSupply: ", totalSupply);
        console.log("wethValue: ", wethValue);
        console.log("wbtcValue: ", wbtcValue);

        assert(wethValue + wbtcValue >= totalSupply);
    }
}
```

The error with -vvvvv

```console
Ran 1 test for test/fuzz/OpenInvariantsTests.t.sol:InvaritantTest
[FAIL. Reason: failed to set up invariant testing environment: No contracts to fuzz.] invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 0, calls: 0, reverts: 0)
Traces:
  [9156033] InvaritantTest::setUp()
    ├─ [4107198] → new <unknown>@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 20402 bytes of code
    ├─ [4849134] 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f::run()
    │   ├─ [2987218] → new <unknown>@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [372255] → new <unknown>@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1082 bytes of code
    │   │   ├─ [449376] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← [Return] 2019 bytes of code
    │   │   ├─ [372255] → new <unknown>@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← [Return] 1082 bytes of code
    │   │   ├─ [449376] → new <unknown>@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   └─ ← [Return] 2019 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 5385 bytes of code
    │   ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← [Return] 
    │   ├─ [591455] → new <unknown>@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← [Return] 2610 bytes of code
    │   ├─ [1030442] → new <unknown>@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← [Return] 4469 bytes of code
    │   ├─ [2424] 0x5FbDB2315678afecb367f032d93F642f64180aa3::transferOwnership(0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512)
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512)
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] 0x5FbDB2315678afecb367f032d93F642f64180aa3, 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 0x104fBc016F4bb334D775a19E8A6510109AC63E00
    ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    └─ ← [Stop] 

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 2.29ms (922.23µs CPU time)

Ran 1 test suite in 651.76ms (2.29ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/fuzz/OpenInvariantsTests.t.sol:InvaritantTest
[FAIL. Reason: failed to set up invariant testing environment: No contracts to fuzz.] invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 0, calls: 0, reverts: 0)

Encountered a total of 1 failing tests, 0 tests succeeded
```

**Reply by EngrPips** [reply]
run `foundryup` then `forge clean` then retry

**Reply by JoshuyPan** [reply]
thanks buddy!

**Reply by EngrPips** [reply]
You are welcome, Friend. Keep crushing it.

**Reply by smilingkylan** [reply]
@EngrPips That worked for me too. Does that mean something just got cached when it wasn't supposed to?

**Reply by EngrPips** [reply]
Hey @kylanhurt, Yeah, that is most likely the case.
